File: ts2mls/js/src/test/scala/Optional.scala
import org.scalatest.funsuite.AnyFunSuite
import ts2mls.TSProgram
import ts2mls.types._

class Optional extends AnyFunSuite {
  test("Optional") {
    val program = TSProgram(Seq("src/test/typescript/Optional.ts"))
    assert(TypeCompare(program.>("buildName"), "(string, string | undefined) => string"))
    assert(TypeCompare(program.>("buildName2"), "(string, string | undefined) => string"))
    assert(TypeCompare(program.>("buildName3"), "(string, any[]) => string"))

    assert(TypeCompare(program.>("SquareConfig").>("color"), "string | undefined"))
    assert(TypeCompare(program.>("SquareConfig").>("width"), "number | undefined"))

    assert(TypeCompare(program.>("did"), "(number, ((number) => number) | undefined) => number"))
    assert(TypeCompare(program.>("getOrElse"), "(object[] | undefined) => object"))
    assert(TypeCompare(program.>("testABC"), "(ABC | undefined) => void"))
    assert(TypeCompare(program.>("testSquareConfig"), "(SquareConfig | undefined) => void"))
    assert(TypeCompare(program.>("err"), "([number, string] | undefined) => void"))
    assert(TypeCompare(program.>("toStr"), "(number | boolean | undefined) => string"))
    assert(TypeCompare(program.>("boo"), "(T' & U' | undefined) => void"))
    assert(TypeCompare(program.>("boom"), "(B<never> | undefined) => any"))
  }
}

File: ts2mls/js/src/test/typescript/Optional.ts
function buildName(firstName: string, lastName?: string) {
  return firstName + lastName;
}

function buildName2(firstName: string, lastName = "DIO") {
  return firstName + lastName;
}

function buildName3(firstName: string, ...lastName) {
  console.log(lastName)
  return firstName;
}

interface SquareConfig {
  color?: string;
  width?: number;
}

function did(x: number, f?: (x: number) => number): number {
  if (f === undefined) return x;
  else return f(x);
}

function getOrElse(arr?: object[]): object {
  if (arr === undefined) return {};
  else return arr[0];
}

class ABC {}
function testABC(abc?: ABC) {}

function testSquareConfig(conf?: SquareConfig) {}

function err(msg?: [number, string]): void {
  if (msg === undefined) return;
  else throw msg;
}

function toStr(x?: number | boolean): string {
  if (x === undefined) return "undefined";
  else return x.toString();
}

function boo<T, U>(x?: T & U) {}

class B<T> {}

function boom(b?: B<never>): any {}